### PR TITLE
14.1.6 support

### PR DESCRIPTION
**Pull request by as3boyan**
14.1.6 support

### CODE REVIEW

File: src/14.1/com/intellij/plugins/haxe/ide/formatter/HaxeFormattingModel.java [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+    return myModel.getDocumentModel();
+  }
+
+  public TextRange replaceWhiteSpace(TextRange textRange, String whiteSpace) {
+    return myModel.replaceWhiteSpace(textRange, whiteSpace);
+  }
+
+  @Override
+  public TextRange shiftIndentInsideRange(ASTNode node, TextRange range, int i) {
+    return myModel.shiftIndentInsideRange(node, range, i);
```
Comment (as3boyan) [inline src/14.1/com/intellij/plugins/haxe/ide/formatter/HaxeFormattingModel.java]: This line was changed, API changes

File: src/14/com/intellij/plugins/haxe/ide/formatter/HaxeFormattingModel.java [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+  public FormattingDocumentModel getDocumentModel() {
+    return myModel.getDocumentModel();
+  }
+
+  public TextRange replaceWhiteSpace(TextRange textRange, String whiteSpace) {
+    return myModel.replaceWhiteSpace(textRange, whiteSpace);
+  }
+
+  public TextRange shiftIndentInsideRange(TextRange range, int indent) {
+    return myModel.shiftIndentInsideRange(range, indent);
```
Comment (as3boyan) [inline src/14/com/intellij/plugins/haxe/ide/formatter/HaxeFormattingModel.java]: This won't work in 14.1.6

**Comment by as3boyan** [comment]
All tests are passed in 14.1.6 except

**Comment by EBatTiVo** [comment]
Because there are no structural changes here, another way to do this is to leave the file in the original location, but add a compile-time constant to call the different API:

In IdeaTarget.java:

<pre>
  public static final int SDK_14_1_CODELINE = 141
  public static final int SDK_14_1_6_BUILD_NUMBER = 3056
  public static final boolean IS_VERSION_14_1_6_OR_LATER =
                                   (SDK_CODELINE >= SDK_14_1_CODELINE 
                                    && SDK_BUILD_NUMBER >= SDK_14_1_6_BUILD_NUMBER)
</pre>

And in the sources:
<pre>
  if (IdeaTarget.IS_VERSION_14_1_6_OR_LATER) {
    // use new api
  } else {
    // use old api
  }
</pre>

**Comment by EBatTiVo** [comment]
BTW, all tests were passing in Linux.

**Comment by as3boyan** [comment]
I don't really understand, override function API was changed for `shiftIndentInsideRange`, I can't see the way how it could be implemented

**Comment by EBatTiVo** [comment]
Can you add the old function signature back to the object?  I think so.  But maybe it's harder than I think.

Once we get clean builds, I'll take the change and I'll see about re-working it the way I envision.  I want to get rid of the v13/v14/v15 source trees as much as possible.  They're already becoming a maintenance headache, since I have to make changes in several files any time I touch one of them.

**Comment by as3boyan** [comment]
old function signature won't compile, because overridden function doesn't exists, 
Their FormattingModel class in IDEA was changed

>I want to get rid of the v13/v14/v15 source trees as much as possible. They're already becoming a maintenance headache, since I have to make changes in several files any time I touch one of them.
Yeah I agree

**Comment by EBatTiVo** [comment]
It looks like JetBrains has locked down or removed the 14.1.4 version from their distribution site so Travis can't get to it.  We have to find another solution for serving the older builds.  And, we have to add a 14.1.6 build, since it's not compatible with 14.1.4.

**Comment by as3boyan** [comment]
Do you want to keep compatibility with 14.1.4?

**Comment by EBatTiVo** [comment]
> Do you want to keep compatibility with 14.1.4?

Yes, I want to keep compatibility with all 14.x releases.  At least until we get to version 16.

**Comment by EBatTiVo** [comment]
By the way, the breaking change was released in 14.1.5, not 14.1.6.

**Comment by as3boyan** [comment]
So version number should be 2735

**Comment by EBatTiVo** [comment]
> So version number should be 2735

Yes, and the names of the properties files should contain 14.1.5 instead of 14,1,6.  And the 14.1 properties file needs to be limited to builds 1532 (14.1.4) or less.  But just in case folks use intermediate builds, we can be nice and limit 14.1 at 2734 (unless you want to go find the actual build where it changed, but that's a bit of work, I think.

**Comment by EBatTiVo** [comment]
This PR was superseded by #384.  Closing.

**Comment by EBatTiVo** [comment]
@as3boyan Any reason to keep the 14.1.6 branch?

**Comment by as3boyan** [comment]
@EBatTiVo I think I moved files from 14.1.6 to 14.1.5, to me there is no reason for keeping separate branch for 14.1.6

**Pull request closed without merging**